### ISSUE TITLE
CI: K8sDatapathConfig: Pods are not ready after timeout: There are some pods with filter that are marked to be deleted

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#7495]
PR: #7476 

Failure seems unrelated to PR contents. Jenkins link:

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/11018/testReport/junit/k8s-1/14/K8sDatapathConfig_Encapsulation_Check_connectivity_with_VXLAN_encapsulation/

### k8s-1.14.K8sDatapathConfig Encapsulation Check connectivity with VXLAN encapsulation

```
There are some pods with filter  that are marked to be deleted
```

#### Stacktrace

```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:409
Pods are not ready after timeout
Expected
    <*errors.errorString | 0xc42428fe70>: {
        s: "There are some pods with filter  that are marked to be deleted",
    }
to be nil
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/DatapathConfiguration.go:110
```

#### Standard Output

```
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Cilium pods: [cilium-rc69z cilium-zkshs]
Netpols loaded: 
CiliumNetworkPolicies loaded: 
Endpoint Policy Enforcement:
Pod                               Ingress   Egress
testclient-4ffph                            
testds-57m8v                      false     false
cilium-etcd-cd4g6klfmm            false     false
cilium-etcd-nvqq2h2hjr            false     false
cilium-operator-6d46676c4-wz85b   false     false
testclient-gnfzv                  false     false
testds-wwkbc                                
cilium-etcd-rlvrglgncx            false     false
coredns-78558db756-ktqjw          false     false
etcd-operator-764667d967-qrt4x    false     false
Cilium agent 'cilium-rc69z': Status: Ok  Health: Ok Nodes "k8s1 k8s2" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 44 Failed 0
Cilium agent 'cilium-zkshs': Status: Ok  Health: Ok Nodes "k8s1 k8s2" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 28 Failed 0
```

#### Standard Error

```
STEP: Performing Cilium preflight check
Cilium is not ready yet: connectivty health is failing: Cluster connectivity is unhealthy on 'cilium-rc69z': Exitcode: 255 
Stdout:
 	 
Stderr:
 	 Error: Cannot get status/probe: Put http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe: dial unix /var/run/cilium/health.sock: connect: no such file or directory
	 
	 command terminated with exit code 255
	 

Cilium is not ready yet: cilium services are not set up correctly: Error validating Cilium service on pod {cilium-rc69z [{0xc422864660 0xc42000e040} {0xc422864720 0xc42000e048} {0xc4228647e0 0xc42000e058} {0xc4228648a0 0xc42000e060} {0xc422864960 0xc42000e070} {0xc422864a20 0xc42000e078}] map[10.109.105.205:2379:[10.10.1.51:2379 (4) 10.10.1.79:2379 (4) 10.10.0.127:2379 (4) 10.10.1.51:2379 (4) 0.0.0.0:0 (0) 10.10.1.79:2379 (4)] 10.96.0.10:53:[0.0.0.0:0 (0) 10.10.1.139:53 (3)] 10.96.0.10:9153:[0.0.0.0:0 (0) 10.10.1.139:9153 (2)] 10.96.0.1:443:[0.0.0.0:0 (0) 192.168.36.11:6443 (1)] 10.97.199.20:10080:[10.10.0.89:80 (6) 10.10.1.8:80 (6) 0.0.0.0:0 (0)] 10.99.46.42:80:[0.0.0.0:0 (0) 10.10.1.8:80 (5) 10.10.0.89:80 (5)]]}: Could not match cilium service backend address 10.10.0.89:80 with k8s endpoint

STEP: Waiting for all etcd-operator pods are ready
=== Test Finished at 2019-03-25T07:48:44Z====
===================== TEST FAILED =====================

cmd: kubectl get pods -o wide --all-namespaces
Exitcode: 0 
Stdout:
 	 NAMESPACE     NAME                                    READY   STATUS        RESTARTS   AGE     IP              NODE   NOMINATED NODE   READINESS GATES
	 default       spaceship-786cbdfbcd-rhmrp              0/1     Terminating   0          45s     10.10.0.236     k8s1   <none>           <none>
	 default       spaceship-786cbdfbcd-smlc7              0/1     Terminating   0          45s     10.10.0.221     k8s1   <none>           <none>
	 default       testclient-4ffph                        1/1     Running       0          26s     10.10.1.63      k8s2   <none>           <none>
	 default       testclient-gnfzv                        1/1     Running       0          26s     10.10.0.37      k8s1   <none>           <none>
	 default       testds-57m8v                            1/1     Running       0          26s     10.10.1.8       k8s2   <none>           <none>
	 default       testds-wwkbc                            1/1     Running       0          26s     10.10.0.89      k8s1   <none>           <none>
	 default       xwing-5d6d4b55cf-pd97n                  0/1     Terminating   0          44s     10.10.0.71      k8s1   <none>           <none>
	 kube-system   cilium-etcd-cd4g6klfmm                  1/1     Running       0          5m56s   10.10.0.127     k8s1   <none>           <none>
	 kube-system   cilium-etcd-nvqq2h2hjr                  1/1     Running       0          5m39s   10.10.1.79      k8s2   <none>           <none>
	 kube-system   cilium-etcd-operator-7ff79779db-g2qzg   1/1     Running       0          6m32s   192.168.36.12   k8s2   <none>           <none>
	 kube-system   cilium-etcd-rlvrglgncx                  1/1     Running       0          6m13s   10.10.1.51      k8s2   <none>           <none>
	 kube-system   cilium-operator-6d46676c4-wz85b         1/1     Running       0          6m29s   10.10.1.59      k8s2   <none>           <none>
	 kube-system   cilium-rc69z                            1/1     Running       0          24s     192.168.36.12   k8s2   <none>           <none>
	 kube-system   cilium-zkshs                            1/1     Running       0          24s     192.168.36.11   k8s1   <none>           <none>
	 kube-system   coredns-78558db756-ktqjw                1/1     Running       0          4m55s   10.10.1.139     k8s2   <none>           <none>
	 kube-system   etcd-k8s1                               1/1     Running       0          11m     192.168.36.11   k8s1   <none>           <none>
	 kube-system   etcd-operator-764667d967-qrt4x          1/1     Running       0          6m27s   10.10.1.73      k8s2   <none>           <none>
	 kube-system   kube-apiserver-k8s1                     1/1     Running       0          11m     192.168.36.11   k8s1   <none>           <none>
	 kube-system   kube-controller-manager-k8s1            1/1     Running       0          11m     192.168.36.11   k8s1   <none>           <none>
	 kube-system   kube-proxy-n84n6                        1/1     Running       0          6m54s   192.168.36.12   k8s2   <none>           <none>
	 kube-system   kube-proxy-nmgzz                        1/1     Running       0          12m     192.168.36.11   k8s1   <none>           <none>
	 kube-system   kube-scheduler-k8s1                     1/1     Running       0          11m     192.168.36.11   k8s1   <none>           <none>
	 
Stderr:
 	 

cmd: kubectl exec -n kube-system cilium-rc69z -- cilium bpf tunnel list
Exitcode: 0 
Stdout:
 	 TUNNEL              VALUE             
	 f00d::a0a:0:0:0:0   192.168.36.11:0   
	 10.10.0.0:0         192.168.36.11:0   
	 
Stderr:
 	 

cmd: kubectl exec -n kube-system cilium-rc69z -- cilium endpoint list
Exitcode: 0 
Stdout:
 	 ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                               IPv6                   IPv4          STATUS   
	            ENFORCEMENT        ENFORCEMENT                                                                                                                 
	 384        Disabled           Disabled          101        k8s:app=etcd                                              f00d::a0a:100:0:6e42   10.10.1.51    ready   
	                                                            k8s:etcd_cluster=cilium-etcd                                                                           
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                               
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                                        
	                                                            k8s:io.cilium/app=etcd-operator                                                                        
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                                            
	 723        Disabled           Disabled          101        k8s:app=etcd                                              f00d::a0a:100:0:ca50   10.10.1.79    ready   
	                                                            k8s:etcd_cluster=cilium-etcd                                                                           
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                               
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                                        
	                                                            k8s:io.cilium/app=etcd-operator                                                                        
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                                            
	 894        Disabled           Disabled          104        k8s:io.cilium.k8s.policy.cluster=default                  f00d::a0a:100:0:7b48   10.10.1.139   ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=coredns                                                        
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                                            
	                                                            k8s:k8s-app=kube-dns                                                                                   
	 1227       Disabled           Disabled          35996      k8s:io.cilium.k8s.policy.cluster=default                  f00d::a0a:100:0:a53c   10.10.1.63    ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                                        
	                                                            k8s:io.kubernetes.pod.namespace=default                                                                
	                                                            k8s:zgroup=testDSClient                                                                                
	 1242       Disabled           Disabled          4          reserved:health                                           f00d::a0a:100:0:127d   10.10.1.205   ready   
	 1707       Disabled           Disabled          105        k8s:io.cilium.k8s.policy.cluster=default                  f00d::a0a:100:0:e085   10.10.1.59    ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=cilium-operator                                                
	                                                            k8s:io.cilium/app=operator                                                                             
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                                            
	                                                            k8s:name=cilium-operator                                                                               
	 2065       Disabled           Disabled          40069      k8s:io.cilium.k8s.policy.cluster=default                  f00d::a0a:100:0:a50b   10.10.1.8     ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                                        
	                                                            k8s:io.kubernetes.pod.namespace=default                                                                
	                                                            k8s:zgroup=testDS                                                                                      
	 3999       Disabled           Disabled          100        k8s:io.cilium.k8s.policy.cluster=default                  f00d::a0a:100:0:1503   10.10.1.73    ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=cilium-etcd-sa                                                 
	                                                            k8s:io.cilium/app=etcd-operator                                                                        
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                                            
	 
Stderr:
 	 

cmd: kubectl exec -n kube-system cilium-zkshs -- cilium bpf tunnel list
Exitcode: 0 
Stdout:
 	 TUNNEL                VALUE             
	 f00d::a0a:100:0:0:0   192.168.36.12:0   
	 10.10.1.0:0           192.168.36.12:0   
	 
Stderr:
 	 

cmd: kubectl exec -n kube-system cilium-zkshs -- cilium endpoint list
Exitcode: 0 
Stdout:
 	 ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                       IPv6                 IPv4          STATUS   
	            ENFORCEMENT        ENFORCEMENT                                                                                                       
	 1234       Disabled           Disabled          4          reserved:health                                   f00d::a0a:0:0:3f6d   10.10.0.27    ready   
	 1825       Disabled           Disabled          101        k8s:app=etcd                                      f00d::a0a:0:0:10c4   10.10.0.127   ready   
	                                                            k8s:etcd_cluster=cilium-etcd                                                                 
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                     
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                              
	                                                            k8s:io.cilium/app=etcd-operator                                                              
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                                  
	 2504       Disabled           Disabled          35996      k8s:io.cilium.k8s.policy.cluster=default          f00d::a0a:0:0:d6cd   10.10.0.37    ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                              
	                                                            k8s:io.kubernetes.pod.namespace=default                                                      
	                                                            k8s:zgroup=testDSClient                                                                      
	 3741       Disabled           Disabled          40069      k8s:io.cilium.k8s.policy.cluster=default          f00d::a0a:0:0:e455   10.10.0.89    ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                              
	                                                            k8s:io.kubernetes.pod.namespace=default                                                      
	                                                            k8s:zgroup=testDS                                                                            
	 
Stderr:
```

[164cd3cc_K8sDatapathConfig_Encapsulation_Check_connectivity_with_VXLAN_encapsulation.zip](https://github.com/cilium/cilium/files/3005477/164cd3cc_K8sDatapathConfig_Encapsulation_Check_connectivity_with_VXLAN_encapsulation.zip)

Possible similarity to https://github.com/cilium/cilium/issues/5256, where a previous test did not clean up services after itself.

**Comment by joestringer** [comment]
Random Seed: 1553499730

**Comment by joestringer** [comment]
Previous test is:
```
07:47:51  K8sDemosTest 
07:47:51    Tests Star Wars Demo
```

Current test:

```
07:48:12  • [SLOW TEST:20.909 seconds]
07:48:12  K8sDemosTest
07:48:12  /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:377
07:48:12    Tests Star Wars Demo
07:48:12    /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:409
07:48:12  ------------------------------
07:48:12  K8sDatapathConfig Encapsulation 
07:48:12    Check connectivity with VXLAN encapsulation
07:48:12    /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:409
07:48:34  STEP: Performing Cilium preflight check
07:48:34  Cilium is not ready yet: connectivty health is failing: Cluster connectivity is unhealthy on 'cilium-rc69z': Exitcode: 255 
07:48:34  Stdout:
07:48:34   	 
07:48:34  Stderr:
07:48:34   	 Error: Cannot get status/probe: Put http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe: dial unix /var/run/cilium/health.sock: connect: no such file or directory
07:48:34  	 
07:48:34  	 command terminated with exit code 255
07:48:34  	 
07:48:34  
07:48:40  Cilium is not ready yet: cilium services are not set up correctly: Error validating Cilium service on pod {cilium-rc69z [{0xc422864660 0xc42000e040} {0xc422864720 0xc42000e048} {0xc4228647e0 0xc42000e058} {0xc4228648a0 0xc42000e060} {0xc422864960 0xc42000e070} {0xc422864a20 0xc42000e078}] map[10.109.105.205:2379:[10.10.1.51:2379 (4) 10.10.1.79:2379 (4) 10.10.0.127:2379 (4) 10.10.1.51:2379 (4) 0.0.0.0:0 (0) 10.10.1.79:2379 (4)] 10.96.0.10:53:[0.0.0.0:0 (0) 10.10.1.139:53 (3)] 10.96.0.10:9153:[0.0.0.0:0 (0) 10.10.1.139:9153 (2)] 10.96.0.1:443:[0.0.0.0:0 (0) 192.168.36.11:6443 (1)] 10.97.199.20:10080:[10.10.0.89:80 (6) 10.10.1.8:80 (6) 0.0.0.0:0 (0)] 10.99.46.42:80:[0.0.0.0:0 (0) 10.10.1.8:80 (5) 10.10.0.89:80 (5)]]}: Could not match cilium service backend address 10.10.0.89:80 with k8s endpoint
07:48:45  STEP: Waiting for all etcd-operator pods are ready
07:48:45  === Test Finished at 2019-03-25T07:48:44Z====
07:48:45  ===================== TEST FAILED =====================
```

Not sure why the preflight checks failed within ~33 seconds of test start, given that the full timeout should be 5 minutes and I would have expected the CI to continue attempting to probe status until it times out:

https://github.com/cilium/cilium/blob/f2dc3771f868848cf9a07290f9df0454a970128b/test/helpers/kubectl.go#L1872-L1891

https://github.com/cilium/cilium/blob/f2dc3771f868848cf9a07290f9df0454a970128b/test/helpers/kubectl.go#L1847-L1850

**Comment by joestringer** [comment]
Oh, the answer is that the preflight actually passes soon after and we actually fail in `STEP: Waiting for all etcd-operator pods are ready`.

**Comment by joestringer** [comment]
The actual error here is a timeout because of:

```
s: "There are some pods with filter that are marked to be deleted",
```

Isn't there supposed to be a filter here? Or is this part of etcd-operator setup in the helpers just expecting that preflight checks have waited until all previous pods were deleted?

**Comment by joestringer** [comment]
OK, it's because the DatapathConfig tests are redeploying too soon after the star wars test ran, and it doesn't explicitly wait for the old pods to be deleted, but then in `WaitForNPods()` it fails out immediately if there are any pods currently being deleted:

https://github.com/cilium/cilium/blob/f2dc3771f868848cf9a07290f9df0454a970128b/test/k8sT/DatapathConfiguration.go#L74-L87

...

https://github.com/cilium/cilium/blob/f2dc3771f868848cf9a07290f9df0454a970128b/test/helpers/kubectl.go#L564-L572

(From the CI failure output:)

```
...
 	 NAMESPACE     NAME                                    READY   STATUS        RESTARTS   AGE     IP              NODE   NOMINATED NODE   READINESS GATES
	 default       spaceship-786cbdfbcd-rhmrp              0/1     Terminating   0          45s     10.10.0.236     k8s1   <none>           <none>
	 default       spaceship-786cbdfbcd-smlc7              0/1     Terminating   0          45s     10.10.0.221     k8s1   <none>           <none>
	 default       testclient-4ffph                        1/1     Running       0          26s     10.10.1.63      k8s2   <none>           <none>
	 default       testclient-gnfzv                        1/1     Running       0          26s     10.10.0.37      k8s1   <none>           <none>
	 default       testds-57m8v                            1/1     Running       0          26s     10.10.1.8       k8s2   <none>           <none>
	 default       testds-wwkbc                            1/1     Running       0          26s     10.10.0.89      k8s1   <none>           <none>
	 default       xwing-5d6d4b55cf-pd97n                  0/1     Terminating   0          44s     10.10.0.71      k8s1   <none>           <none>
...
```

**Comment by joestringer** [comment]
I believe that this should be fixed by #7537.
